File: island_count/island_count.py
​
# set to track visited nodes
# count variable
# dfs recursively to check neighbors
​
# O(nm) time and O(nm) space
def island_count(grid):
  count = 0
  visited = set()
  for i in range(len(grid)):
    for j in range(len(grid[0])):
      if dfs(grid, visited, i, j):
        count += 1
      
  return count
​
def dfs(grid, visited, row, col):
  row_in_bounds = 0 <= row < len(grid)
  col_in_bounds = 0 <= col < len(grid[0])
  if not row_in_bounds or not col_in_bounds:
    return False
  
  if grid[row][col] == 'W':
    return False
  
  position = (row, col)
  if position in visited:
    return False
  visited.add(position)
  
  dfs(grid, visited, row - 1, col)
  dfs(grid, visited, row + 1, col)
  dfs(grid, visited, row, col - 1)
  dfs(grid, visited, row, col + 1)
  
  return True

File: Graph-I/089-island-count/island-count.js
//O (rc) time and O(rc) space, r = rows and c = columns
const islandCount = (grid) => {
  const visited = new Set();
  let count = 0;

  for (let row = 0; row < grid.length; row++) {
    for (let col = 0; col < grid[0].length; col++) {
      if (exploreIsland(grid, row, col, visited) === true) count++;
    }
  }

  return count;
};

const exploreIsland = (grid, row, col, visited) => {
  const rowInbounds = 0 <= row && row < grid.length;
  const colInbounds = 0 <= col && col < grid[0].length;
  if (!rowInbounds || !colInbounds) return false;
  
  const pos = `${row}-${col}`;
  if (visited.has(pos)) return false;
  visited.add(pos);
  
  if (grid[row][col] === 'W') return false;
 
  exploreIsland(grid, row - 1, col, visited);  
  exploreIsland(grid, row + 1, col, visited);  
  exploreIsland(grid, row, col - 1, visited);  
  exploreIsland(grid, row, col + 1, visited);  

  return true;
};


module.exports = {
  islandCount,
};
